### ISSUE TITLE
auth stuck at the callback URL — not storing Auth0 token [SPA mode]

### ISSUE THREAD

**Issue by maxiride** [nuxt-community/auth-module#536]
**EDIT**
All the testing links are not working anymore. Given the feedback and the PR made it _seems_ the issue is solved so I took offline the demos I made for the various scenarios. I will check out the new PR too as soon as I can but given that the project maintainers already merged it to master I suppose it's a go for production too.

### Given the three tests I've done, I am making here a recap for clarity

1. Mode **SPA** — **example-auth0** repository https://github.com/nuxt-community/auth-module/issues/536#issue-568235358
Reproduction link: [https://nuxt.federicod.dev/](https://nuxt.federicod.dev/)

Application stays stuck at the callback page, not storing the auth0 token, in the chance that this was an issue with the example-auth0 repo alone I've tested also the example build in the auth-module repo.

2. Mode **SPA** — **auth-module** repository https://github.com/nuxt-community/auth-module/issues/536#issuecomment-589557690
Reproduction link: https://auth-test.federicod.dev/

Same as above, I now think that the issue resides indeed in the auth-module itself.

3. Mode **SSR** (universal) — **auth-module** repository https://github.com/nuxt-community/auth-module/issues/536#issuecomment-589600857
Reproduction link: https://auth-test-ssr.federicod.dev/

After trying to build the SSR version of the example\demo application, and serving it with `nuxt start` I have confirmed that the auth-module either:

* doesn't work at all in SPA mode;
* the build command with the SPA configuration breaks something
---

### Version

[v4.8.5](https://github.com/auth-module/releases/tag/v4.8.5)

### Reproduction link

[https://nuxt.federicod.dev/](https://nuxt.federicod.dev/)

### Steps to reproduce

1. `git clone https://github.com/nuxt/example-auth0.git`;
2. set Auth0  `domain` and `client_id` into `nuxt.config.js` (I am not using the .env file as it seems to me that it doesn't get embedded in the SPA after running npm run build)
3. set mode: `spa`
4. nuxt build
5. deploy the dist folder
6. access the website and try to login with demo@demo.com; Demo1@23
7. after the callback URL is opened, it stays stuck there

On a more complicated platform I am developing, some API calls that are made on the callback URL to populate a Vuex store all yields unauthorized error, meaning that the token doesn't get stored at all.

When running example-auth0 and the application I am developing in development mode with  `npm run dev`, everything works fine. So I suppose that something goes wrong only after running the build command.

### What is expected ?

I expect the callback URL to be opened and the token received stored.

### What is actually happening?

The SPA stays stuck at the callback URL and no token is stored.

### Additional comments?

Webserver: Caddy
Configuration:

```
nuxt.federicod.dev {
        root /root/dev/example-auth0/dist
}
```

No errors are shown in the console.

<div align="right"><sub><em>This bug report is available on <a href="https://cmty.app/nuxt">Nuxt</a> community (<a href="https://cmty.app/nuxt/auth-module/issues/c500">#c500</a>)</em></sub></div>

**Comment by jas777** [comment]
I have the exact same issue just with custom OAuth routing. If you'll happen to find a solution please mention me. Thanks in advance ^^

**Comment by Maxiride** [comment]
To ensure it wasn't an issue only on `example-auth0` end I've also tried the following with the auth-module repo:

**Reproduction link**
https://auth-test.federicod.dev/

**Steps to reproduce**

1. `git clone https://github.com/nuxt-community/auth-module.git`
1. change domain and client_id in `examples/demo/nuxt.config.js` with my own Auth0 tenant
1. Run `nuxt build examples\demo --spa`
1. Deploy generated dist folder
1. access the website and try to login with demo@demo.com; Demo1@23
1. after the callback URL is opened, it stays stuck there

Same as above, when running auth-module development mode with `nuxt`, everything works fine. So I suppose that something goes wrong only after running the build command.

**Comment by Maxiride** [comment]
### Mode SSR (universal) — auth-module repository

Steps 1 and 2 same as https://github.com/nuxt-community/auth-module/issues/536#issuecomment-589557690

3. Run `nuxt build examples\demo`  `nuxt start examples\demo`
4. **Application works as expected**

I'm making separate comments in the hope to more clearly separate the different tests I've done.

I am now prone to believe that either:

* the build command with the spa flag (or with mode: `spa` configuration) breaks the auth-module;
* the auth-module simply isn't compatible at all with a `spa` configuration

**Comment by Maxiride** [comment]
After a whole day researching I stumbled upon these issues https://github.com/nuxt-community/auth-module/issues/299 and https://github.com/nuxt/nuxt.js/issues/5267.

The whole situation is still somewhat unclear to me, however I've tried to deploy the application in another way again **Mode SSR (spa) — auth-module repository** and the authentication flow worked as expected.

As referenced in the above mentioned issues it seems that static SPA mode is not supported **at all** by the auth-module (more specifically by its dynamic behaviour, so other libraries could be affected). What it is not explain at all in the [documentation](https://nuxtjs.org/guide/commands#single-page-application-deployment-spa-) is that there are two possible ways to serve an SPA application:

- either with the `mode: 'spa'` `nuxt build` and `nuxt start`
- or the `mode: 'spa'` `nuxt build` and directly serving the dist folder as the document root with the webserver in use

In either case, `nuxt build examples\demo` creates two relevant folders:

- the `/dist` folder in the project root;
- the `/example/demo/.nuxt`;

```
.
|-- dist
|   |-- _nuxt
|   |   `-- img
|   |-- callback
|   |-- login
|   |-- public
|   `-- secure
|-- examples
|   |-- api
|   |-- demo
|   |   |-- .nuxt
|   |   |-- assets
|   |   |-- components
|   |   |-- layouts
|   |   |-- pages
|   |   `-- store
```

From the [documentation on SPA](https://nuxtjs.org/guide/commands#single-page-application-deployment-spa-) one would assume that deploying only the `/dist` folder is necessary, instead to have the authentication flow working properly it is needed to run `nuxt start examples/demo` which serves the files from `/example/demo/.nuxt`.

I am hella lot confused on the deployment process involved with Nuxt. I beg someone to explain what's happening here.

**Comment by arambert** [comment]
After some tests on my side it seems to me that on SPA mode, on initial load (first page or hard refresh), the route (and maybe some other elements) are not properly initialized when passed to the middleware.
for instance ctx.route.matched seems to be always empty in the middlewares for the first page (SPA mode), even if there is a page matching.
This breaks auth-module on first pages on SPA mode (the callback from auth0 is a first page...)
I wonder if other have noticed the same issue.

**Comment by krutijan1** [comment]
Its not working for me either, netlify + auth0 mode spa, on local it's working, deployed generated version not

**Comment by Maxiride** [comment]
No clue either, but check https://github.com/nuxt-community/auth-module/issues/299#issuecomment-591282658

**Comment by vedovelli** [comment]
Same here. Nuxt deployed on Netlify. 

Redirect occurs (after adding trailing slash to callback) but in the store all relevant data for auth is null.

**Comment by vedovelli** [comment]
My solution was to move my app from Netlify to Heroku. I changed the mode to **universal** and now although out of my favorite hosting company, I benefit from SSR. =) And the auth module is working like a charm.

**Comment by arambert** [comment]
The Nuxt community has a strong focus on SSR and sometimes the likes of us that need just a SPA can feel a little left behind...

**Comment by skydiver** [comment]
This is my solution to host on Zeit/Now and login using github:

`now.json`
```
{
  "builds": [
    { "src": "package.json", "use": "@now/static-build" },
    { "src": "/functions/*.js", "use": "@now/node" }
  ],
  "routes": [
    { "src": "/_auth/oauth/github/authorize", "methods": ["POST"], "dest": "/functions/authorize.js" }
  ],
  "env": {
    "GITHUB_CLIENT_ID": "YOUR CLIENT ID",
    "GITHUB_CLIENT_SECRET": "YOUR CLIENT SECRET"
  },
  "build": {
    "env": {
      "GITHUB_CLIENT_ID": "YOUR CLIENT ID",
      "GITHUB_CLIENT_SECRET": "YOUR CLIENT SECRET"
    }
  }
}
```

`functions/authorize.js`
```
const axios = require('axios');

require('dotenv').config();

module.exports = (req, res) => {
  const {
    code,
    redirect_uri: redirectUri,
    response_type: responseType
  } = req.body;

  axios
    .request({
      method: 'post',
      url: 'https://github.com/login/oauth/access_token',
      data: {
        client_id: process.env.GITHUB_CLIENT_ID,
        client_secret: process.env.GITHUB_CLIENT_SECRET,
        redirectUri,
        responseType,
        code
      },
      headers: {
        Accept: 'application/json'
      }
    })
    .then((response) => {
      res.status(200).send(response.data);
    })
    .catch((error) => {
      res.status(403).send(error);
    });
};
```

It's just a serverless function who overrides `/_auth/oauth/github/authorize` and perform github authorization.

**Comment by rajatjindal** [comment]
Hi Fellow users

I have submitted a PR with a potential fix https://github.com/nuxt-community/auth-module/pull/586 (it fixed my login flow which has same issue), can some of you please try the fix and confirm if it fixes your issue too (or not). 

Thanks
Rajat Jindal

**Comment by Maxiride** [comment]
> 
> 
> Hi Fellow users
> 
> I have submitted a PR with a potential fix #586 (it fixed my login flow which has same issue), can some of you please try the fix and confirm if it fixes your issue too (or not).
> 
> Thanks
> Rajat Jindal

I was about to test it out but it has already been merged! So I suppose it's go from the devs too!
Will check it out anyway =)

**Comment by rajatjindal** [comment]
thank you. will wait for you to confirm as well. I have also requested to cut a new release with this fix (if its not too much to ask).

**Comment by sugoidesune** [comment]
I am having the same issue with google auth.
The state "token" is being set correctly, but not the access_token.
Curiously, the auth.strategy is set to google on load of the callbackurl and switches after 2 seconds to 'local'.

**Comment by rajatjindal** [comment]
I think for google auth, u need authorize endpoint. One of the comment above talks about setting that for github auth case and also talks about how netlify redirect can solve it

https://github.com/nuxt-community/auth-module/issues/536#issuecomment-600721072

May be that is the issue?

**Comment by sugoidesune** [comment]
I authorized the callback url, no issues from google.
But the auth module is completely non reactive when i arrive at the callback url with the tokens in the url. I have now added the following to my callback url page and it works.

![image](https://user-images.githubusercontent.com/20317691/81709892-8498a400-9472-11ea-86c1-3e17d1808c45.png)
But thats as manual as it can get. Certainly this is something this module should take care of.
at least
